### PR TITLE
PLAT-33: Adding tests for Platform UI

### PR DESCRIPTION
**Pull request by Charpell**
## JIRA TICKET NAME
[PLAT-33: Adding tests for Platform UI ](https://issues.openmrs.org/browse/PLAT-33)

## SUMMARY:
To ensure all the components has all functionality work the way it was intended.
We use Jest to setup platform testing environment.

Keynote:
Set up Jest environment
Setup coverage Report
Setup Jest watch mode

## Issue I worked on
[PLAT-33](https://issues.openmrs.org/browse/PLAT-33)

## Checklist: I completed these to help reviewers
 - [x] My pull request only contains ONE single commit (the number above, next to the 'Commits' tab is 1).
No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [x] My IDE is configured to follow the [code style](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.
No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

 - [ ] I have **added tests** to cover my changes. (If you refactored existing code that was well tested you do not have to add tests)
No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and added all formatting changes to my commit.
No? -> execute above command

- [ ] All new and existing **tests passed**.
No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

 - [x] My pull request is **based on the latest changes** of the master branch.
No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: __tests__/Login.test.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  target: {
+    name: 'name',
+    value: 'value',
+  },
+  preventDefault: () => jest.fn()
+};
+
+
+describe('Login Component', () => {
+  // const { wrapper, props } = setup();
```
Comment (segunolalive) [inline __tests__/Login.test.js]: remove unused code

**Review by aenkya — COMMENTED** [review]
Looks good to me.

I would like to recommend that we use less tests that check number of elements like divs
Read up on Snapshot testing

File: __tests__/Login.test.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    expect(wrapper.find('div').length).toBe(7);
+    expect(wrapper.find('form').length).toBe(1);
+    expect(wrapper.find('h4').length).toBe(1);
+    expect(wrapper.find('label').length).toBe(2);
+    expect(wrapper.find('input').length).toBe(2);
+    expect(wrapper.find('button').length).toBe(1);
+
+  })
+
+  it('should have all the method in the component to be defined', () => {
```
Comment (segunolalive) [inline __tests__/Login.test.js]: this test suite provides no benefits. Nothing is really being tested.

Test instead that the method is called when appropriate and that it functions appropriately

File: package.json [hunk 3]
```diff
@@ -8,39 +8,47 @@
...
     "karma-chrome-launcher": "^2.2.0",
     "karma-coverage": "^1.1.1",
     "karma-firefox-launcher": "^1.0.1",
     "karma-jasmine": "^1.1.0",
     "karma-sourcemap-loader": "^0.3.7",
     "karma-webpack": "^2.0.4",
+    "on-build-webpack": "^0.1.0",
     "raw-loader": "^0.5.1",
+    "react-addons-test-utils": "^15.6.2",
+    "react-dom": "^15.6.2",
```
Comment (segunolalive) [inline package.json]: This should not be a devDependency. It is crucial to production

File: __tests__/Login.test.js (hunk 2 above)
Comment (segunolalive) [inline __tests__/Login.test.js]: this test suite provides no benefits. Nothing is really being tested.

Test instead that the method is called when appropriate and that it functions appropriately

**Review by segunolalive — CHANGES_REQUESTED** [review]
Good job overall

File: __tests__/Login.test.js (hunk 1 above)
Comment (Charpell) [reply __tests__/Login.test.js]: The PR has been updated

File: __tests__/Login.test.js (hunk 2 above)
Comment (Charpell) [reply __tests__/Login.test.js]: The PR has been updated

File: package.json (hunk 3 above)
Comment (Charpell) [reply package.json]: The PR has been updated

File: __tests__/Login.test.js (hunk 2 above)
Comment (Charpell) [reply __tests__/Login.test.js]: The PR has been updated

**Comment by segunolalive** [comment]
LGTM

File: app/js/components/presentational/LoginForm.jsx [hunk 4]
```diff
@@ -9,7 +9,7 @@
 import React from 'react';
```
Comment (patlub) [inline app/js/components/presentational/LoginForm.jsx:9]: @Charpell Is this file `LoginForm.jsx` part of the tests?

Comment (Charpell) [reply app/js/components/presentational/LoginForm.jsx:9]: @patlub Yes

**Comment by dkayiwa** [comment]
@Charpell what do you think of @Elbertbiggs360 's comment?

**Comment by Charpell** [comment]
@dkayiwa Snapshot testing is good but it's not an alternative to traditional tests

**Pull request merged**